compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp

```diff
@@ -40,28 +42,77 @@ static Value allocateGlobalSharedMemory(Location loc, OpBuilder &builder,
   return builder.create<memref::AllocOp>(loc, memrefType);
 }
 
+/// Packs vector of lower precision into a single 32-bit width element.
+/// (i.e <2xf16> -> f32 and <4xi8> -> i32)
+static Value packVectorToSupportedWidth(Location loc, OpBuilder &builder,
+                                        Value input) {
+  auto vecType = input.getType().cast<VectorType>();
+  assert(vecType && "This packing fn only works if input is a vector.");
+  Type elementType = vecType.getElementType();
+  int64_t vecSize = vecType.getShape()[0];
+  unsigned bitWidth = elementType.getIntOrFloatBitWidth();
+  // Check that numel * bitwidth = 32.
+  assert(vecSize * bitWidth == kShuffleBitWidth &&
+         "vector needs to packable into 32-bitwidth.");
+  assert(elementType.isIntOrFloat() && "Only int and float supported.");
+  Type supportedType = builder.getF32Type();
+  if (elementType.dyn_cast<IntegerType>()) {
+    supportedType = builder.getI32Type();
+  }
+  VectorType packed32Type = VectorType::get({1}, supportedType);
+  Value packedInputVec =
+      builder.create<vector::BitCastOp>(loc, packed32Type, input);
+  Value packedInput = builder.create<vector::ExtractOp>(loc, packedInputVec, 0);
+  return packedInput;
+}
+
+static Value promoteElementToVector(Location loc, OpBuilder &builder,
+                                    Value input) {
+  VectorType vectorTypeBroadcast = VectorType::get({1}, input.getType());
+  Value vectorInput =
+      builder.create<vector::BroadcastOp>(loc, vectorTypeBroadcast, input);
+  return vectorInput;
+}
+
 /// Emit warp reduction code sequence for a given input.
 static Value warpReduction(Location loc, OpBuilder &builder, Value input,
                            vector::CombiningKind kind, uint32_t warpSize,
                            uint32_t numLaneToReduce) {
+  VectorType unpackedType = input.getType().dyn_cast<VectorType>();
   Value laneVal = input;
   assert(llvm::isPowerOf2_32(numLaneToReduce));
   // Parallel reduction using butterfly shuffles.
   for (uint64_t i = 1; i < numLaneToReduce; i <<= 1) {
+    Value shuffleInput = laneVal;
+    if (unpackedType) {
+      shuffleInput = packVectorToSupportedWidth(loc, builder, laneVal);
+    }
     Value shuffled = builder
-                         .create<gpu::ShuffleOp>(loc, laneVal, i,
+                         .create<gpu::ShuffleOp>(loc, shuffleInput, i,
                                                  /*width=*/warpSize,
                                                  /*mode=*/gpu::ShuffleMode::XOR)
                          .getShuffleResult();
+    if (unpackedType) {
+      Value vectorShuffled = promoteElementToVector(loc, builder, shuffled);
+      shuffled =
+          builder.create<vector::BitCastOp>(loc, unpackedType, vectorShuffled);
```

Comment:
nit: I would make it a unpackVector symmetrical to `packVectorToSupportedWidth`

Reply:
done!